src/d3a/models/market/two_sided.py

```diff
@@ -305,19 +328,33 @@ def _validate_requirements_satisfied(
 
     @classmethod
     def validate_bid_offer_match(
-            cls, bid: Bid, offer: Offer, clearing_rate: float, selected_energy: float) -> None:
-        """Basic validation function for bid against an offer.
+            cls, bids: List[Bid], offers: List[Offer],
+            clearing_rate: float, selected_energy: float) -> None:
+        """Basic validation function for bids against offers.
 
         Raises:
             InvalidBidOfferPairException: Bid offer pair failed the validation
         """
-        if not (bid.energy >= selected_energy and
-                offer.energy >= selected_energy and
-                (bid.energy_rate + FLOATING_POINT_TOLERANCE) >= clearing_rate and
-                (bid.energy_rate + FLOATING_POINT_TOLERANCE) >= offer.energy_rate):
+        bids_total_energy = sum([bid.energy for bid in bids])
+        offers_total_energy = sum([offer.energy for offer in offers])
+        # All combinations of bids and offers
+        # Example List1: [A, B], List2: [C, D] -> combinations: [(A, C), (A, D), (B, C), (B, D)]
+        bids_offers_combinations = itertools.product(bids, offers)
+        if not (
+                bids_total_energy >= selected_energy and
+                offers_total_energy >= selected_energy
+                and all(
+                    (bid.energy_rate + FLOATING_POINT_TOLERANCE) >= clearing_rate for bid in bids)
+                and all(
+                    combination[0].energy_rate +
+                    FLOATING_POINT_TOLERANCE >= combination[1].energy_rate
```

Comment:
I understand that the 0th element is the bid and the 1st element is the offer, right? Can you please add this to your comment in line 341? Thanks!

Reply:
Furthermore, as far as I see, the reason for the permutation here is in order to validate that all bids have a higher energy rate than all offers, right ? In this case, in order to avoid the product which can potentially be expensive for long lists of matches, why not simply checking that all offers are less than or equal to the clearing rate ? In combination with the check of the bids it will achieve the same result, assuming that the clearing rate is the same for all pairs.